modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/DirectProxyAsmGenerator.java

```diff
@@ -0,0 +1,297 @@
+                arg("changer", DynamicConfigurationChanger.class)
+        );
+
+        //     super(keys, changer);
+        // }
+        ctor.getBody()
+                .append(ctor.getThis())
+                .append(ctor.getScope().getVariable("keys"))
+                .append(ctor.getScope().getVariable("changer"))
+                .invokeConstructor(DIRECT_CFG_CTOR)
+                .ret();
+    }
+
+    /**
+     * Generates getter based on the field.
+     */
+    private void addGetMethod(Field schemaField) {
+        Class<?> schemaFieldType = schemaField.getType();
+
+        String fieldName = schemaField.getName();
+
+        SchemaClassesInfo schemaClassInfo = cgen.schemaInfo(schemaFieldType);
+
+        ParameterizedType returnType;
+
+        // Return type is determined like in ConfigurationImpl class.
+        if (isConfigValue(schemaField)) {
+            returnType = typeFromJavaClassName(schemaClassInfo.cfgClassName);
+        } else if (isNamedConfigValue(schemaField)) {
+            returnType = type(NamedConfigurationTree.class);
+        } else {
+            assert isValue(schemaField) || isPolymorphicId(schemaField) || isInjectedName(schemaField)
+                    || isInternalId(schemaField) : schemaField;
+
+            returnType = type(ConfigurationValue.class);
+        }
+
+        MethodDefinition methodDef = classDef.declareMethod(
+                of(PUBLIC),
+                fieldName,
+                returnType
+        );
+
+        BytecodeBlock body = methodDef.getBody();
+
+        if (isValue(schemaField) || isPolymorphicId(schemaField) || isInjectedName(schemaField) || isInternalId(schemaField)) {
+            // new DirectValueProxy(appendKey(this.keys, new KeyPathNode("name")), changer);
+            // or
+            // new DirectValueProxy(appendKey(this.keys, new KeyPathNode("<internal_id>")), changer);
+            body.append(newInstance(
+                    DirectValueProxy.class,
+                    invokeStatic(
+                            APPEND_KEY,
+                            methodDef.getThis().getField("keys", List.class),
+                            newInstance(
+                                    KeyPathNode.class,
+                                    constantString(isInjectedName(schemaField) ? InnerNode.INJECTED_NAME
+                                            : isInternalId(schemaField) ? InnerNode.INTERNAL_ID : fieldName)
+                            )
+                    ),
+                    methodDef.getThis().getField("changer", DynamicConfigurationChanger.class)
+            ));
+        } else {
+            SchemaClassesInfo fieldSchemaClassInfo = cgen.schemaInfo(schemaField.getType());
+
+            ParameterizedType resultType = typeFromJavaClassName(fieldSchemaClassInfo.directProxyClassName);
+
+            if (isConfigValue(schemaField)) {
+                // new BarDirectProxy(appendKey(this.keys, new KeyPathNode("name")), changer);
```

Comment:
What's Bar?

Reply:
**ParameterizedType resultType = typeFromJavaClassName(fieldSchemaClassInfo.directProxyClassName);**
Example of direct proxy name